### PR TITLE
Use Weight type for TxIn and TxOut

### PR DESCRIPTION
**Pull request by yancyribbens**
Return the `Weight` type for methods in `TxIn` and `TxOut`.

### CODE REVIEW

**Comment by Kixunil** [comment]
The return type is intentionally `usize` because it's error-prone to blindly sum them. (Although `usize` doesn't signal that well...)

Perhaps write a comment about it instead? We could have another newtype that'd act as a lint. Something like:

```rust
/// Weight of transaction component.
///
/// This type acts as a lint to prevent accidental summing of weights without accounting for the length of encoded varint.
pub struct PartialWeight(u64);

impl PartialWeight {
    pub fn to_u64(&self) -> u64 { self.0 }
    /// Sums up multiple weights while accounting for varint length.
    fn sum_with_len(iter: impl Iterator<Item=Self>) -> Weight {
        let mut count = 0;
        Weight(iter.map(|weight| weight.0).inspect(|_| count += 1).sum::<u64>() + VarInt(count).len())
    }
}
```

**Comment by yancyribbens** [comment]
> The return type is intentionally usize because it's error-prone to blindly sum them. 

Can you be more specific about how its error prone?  Like for example what type of error?  Agree a comment made be helpful, I thought returning the Weight type was just not yet implemented.

**Comment by yancyribbens** [comment]
> ..without accounting for the length of encoded varint.

This part I don't understand.

**Comment by tcharding** [comment]
I'll take a stab it answering this. Since weight is based on the wire serialized format, adding inputs and outputs must include the varint used at the front for number of inputs/outputs during consensus serialization (see `consensus_encode` impl on `Vec<T>`).

**Comment by Kixunil** [comment]
> Can you be more specific about how its error prone?

To add to @tcharding  correct explanation, here's naive code that appears to be correct but it isn't:

```rust
fn insert_output(&mut self, output: TxOut) {
    self.total_weight += output.weight;
    self.transaction.output.push(output);
}
```

**Comment by yancyribbens** [comment]
@tcharding @Kixunil thanks, that helps some.  What if all the UTXO's being summed are segwit (does not use legacy_weight)?  Then is is possible to simply sum all `UTXO's` in the "naive" fashion?

**Comment by Kixunil** [comment]
The problem is the same regardless of number of segwit/non-segwit inputs.

**Comment by yancyribbens** [comment]
> The problem is the same regardless of number of segwit/non-segwit inputs.

Ok, I asked because the comment was on the `legacy_weight` method.  I'm not sure what `VarInt` actually is.  Sorry for the dumb questions.  My intuition is that there is some fields in the transaction that can change in size which is why we can't just sum the weights in a naive fashion?

**Comment by Kixunil** [comment]
No problem, this stuff is tricky, even I got just burned.

Yes, the count of inputs, outputs and witnesses is encoded in a way that has different length based on the count, so summing too many inputs will increase the size by more.

We should make sure to have correct comments on all methods. (Or really other types if people aren't too annoyed by having billion types.)

File: bitcoin/src/blockdata/weight.rs [hunk 1]
```diff
@@ -215,6 +215,12 @@ impl From<Weight> for u64 {
     fn from(value: Weight) -> Self { value.to_wu() }
 }
 
+impl Add<u64> for Weight {
+    type Output = Weight;
```
Comment (stevenroose) [inline bitcoin/src/blockdata/weight.rs:219]: Hmm semantically this is wrong right? You should convert the u64 to weight and add that.

Comment (yancyribbens) [reply bitcoin/src/blockdata/weight.rs:219]: Yes I think I see what you mean.  To convert to Weight from u64 we would be explicit about from kwu I think.

**Comment by yancyribbens** [comment]
closing for now since there doesn't seem to be consensus for this.  The current return value `usize` doesn't convey that a transactions weight depends on `VarInt` although this PR doesn't improve that aspect.  Perhaps it's possible for `Vec<TxOut>` to be changed to a custom iterator that has a `sum()` that does the maths correctly.

**Pull request closed without merging**